<a href="https://colab.research.google.com/github/Logan-Stark/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [0]:
### YOUR WORK HERE
# Import different libraries
import pandas as pd 
from scipy import stats
import numpy as np

In [88]:
# Bash command to view raw data
# we can see some '?' we will have to replace those
!curl 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'

3,?,alfa-romero,gas,std,two,convertible,rwd,front,88.60,168.80,64.10,48.80,2548,dohc,four,130,mpfi,3.47,2.68,9.00,111,5000,21,27,13495
3,?,alfa-romero,gas,std,two,convertible,rwd,front,88.60,168.80,64.10,48.80,2548,dohc,four,130,mpfi,3.47,2.68,9.00,111,5000,21,27,16500
1,?,alfa-romero,gas,std,two,hatchback,rwd,front,94.50,171.20,65.50,52.40,2823,ohcv,six,152,mpfi,2.68,3.47,9.00,154,5000,19,26,16500
2,164,audi,gas,std,four,sedan,fwd,front,99.80,176.60,66.20,54.30,2337,ohc,four,109,mpfi,3.19,3.40,10.00,102,5500,24,30,13950
2,164,audi,gas,std,four,sedan,4wd,front,99.40,176.60,66.40,54.30,2824,ohc,five,136,mpfi,3.19,3.40,8.00,115,5500,18,22,17450
2,?,audi,gas,std,two,sedan,fwd,front,99.80,177.30,66.30,53.10,2507,ohc,five,136,mpfi,3.19,3.40,8.50,110,5500,19,25,15250
1,158,audi,gas,std,four,sedan,fwd,front,105.80,192.70,71.40,55.70,2844,ohc,five,136,mpfi,3.19,3.40,8.50,110,5500,19,25,17710
1,?,audi,gas,std,four,wagon,fwd,front,105.80,192.70,71.40,55.70,2954,ohc,five,136,mpfi,3.19,3.40,8.50,1

In [89]:
# read in our data 
names1= ['symboling','normalized-losses','make', 'fuel-type', 'aspiration', 
         'num-of-doors', 'body-style','drive-wheels', 'engine-location', 
         'wheel-base' ,'length' ,'width','height','curb-weight','engine-type' ,
         'num-of-cylinders','engine-size', 'fuel-system', 'bore','stroke', 
         'compression-ratio', 'horsepower','peak-rpm' ,'city-mpg', 
         'highway-mpg','price']     

df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data', names = names1)
print(df.shape)
df.head()

(205, 26)


symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [90]:
# replace '?' with NaNs
df = df. replace({'?':np.nan})
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [91]:
df.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [92]:
### YOUR WORK HERE
# first we need to find the mean of our population
df.describe()  

symboling  wheel-base  ...    city-mpg  highway-mpg
count  205.000000  205.000000  ...  205.000000   205.000000
mean     0.834146   98.756585  ...   25.219512    30.751220
std      1.245307    6.021776  ...    6.542142     6.886443
min     -2.000000   86.600000  ...   13.000000    16.000000
25%      0.000000   94.500000  ...   19.000000    25.000000
50%      1.000000   97.000000  ...   24.000000    30.000000
75%      2.000000  102.400000  ...   30.000000    34.000000
max      3.000000  120.900000  ...   49.000000    54.000000

[8 rows x 10 columns]

 Our $\mu$ is 30.75

In [0]:
# now to make our sample and take our mean from it
sample1 = df['highway-mpg'].sample(n=10, random_state=30)

In [94]:
sample1.mean()

32.8

 Our $\bar{x}$ is 32.8
 This means our $\bar{x}$ is greater than our $\mu$ this is likely 
 because our sample size was too small.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [95]:
### YOUR WORK HERE
stats.ttest_1samp(sample1, popmean=35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

- From our T test we can see that the null hypothesis of " our sample mean will equal the specified $\mu$ given by the salseman" may be true. This is because our p-value is larger than .05. we can also see from our - t test statistic that our $\bar{x}$ is less than the $\mu$

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [0]:
### YOUR WORK HERE


In [97]:
# First we make our sample
carsample2 = df['highway-mpg'].sample(n=100, random_state=30)
carsample2.mean()

30.64

In [98]:
stats.ttest_1samp(carsample2, popmean=35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

- With a sample size of 100 we can see our T test shows a big discrepancy in our T statistic. The bigger our T statistic is than 0 the bigger the discrepancy in our means. our p-value is also much smaller than .05.  This means the $\mu$ must be quite a bit smaller than what the salesman is trying to make us believe. 

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [99]:
### YOUR WORK HERE
# loading our data

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-06-08 23:02:05--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.1’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.03s   

2020-06-08 23:02:05 (632 KB/s) - ‘house-votes-84.data.1’ saved [18171/18171]



In [100]:
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df2 = pd.read_csv('house-votes-84.data', names=column_headers)

print(df.shape)
df.head()

(205, 26)


symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [101]:
df2 = df2.replace({'y': 1, 'n': 0, '?': np.nan})
df2.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [0]:
### YOUR WORK HERE
# First we have to make our two samples for rep and dem
dem = df2[df2['party'] == 'democrat']
rep = df2[df2['party'] == 'republican']

In [103]:
# Now we can simply use . mean to see which columns to do our T tests on
dem.mean()

handicapped-infants     0.604651
water-project           0.502092
budget                  0.888462
physician-fee-freeze    0.054054
el-salvador-aid         0.215686
religious-groups        0.476744
anti-satellite-ban      0.772201
aid-to-contras          0.828897
mx-missile              0.758065
immigration             0.471483
synfuels                0.505882
education               0.144578
right-to-sue            0.289683
crime                   0.350195
duty-free               0.637450
south-africa            0.935135
dtype: float64

In [104]:
rep.mean()

handicapped-infants     0.187879
water-project           0.506757
budget                  0.134146
physician-fee-freeze    0.987879
el-salvador-aid         0.951515
religious-groups        0.897590
anti-satellite-ban      0.240741
aid-to-contras          0.152866
mx-missile              0.115152
immigration             0.557576
synfuels                0.132075
education               0.870968
right-to-sue            0.860759
crime                   0.981366
duty-free               0.089744
south-africa            0.657534
dtype: float64

In [105]:
# First test 'budget' 
stats.ttest_ind(rep['budget'], dem['budget'], nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

- From this test we can see that Democrats supported this bill much more than Republicans. This is easy to see when you look at both the T statistic and p-value. Our T Statistic is farr greater than 0 and it is negative. This would indicate that our mean for 'rep' is much smaller than 'dem'. likewise our miniscule p-value show a big discrepancy in means. 

In [106]:
# Second Test 'physician-fee-freeze'
stats.ttest_ind(rep['physician-fee-freeze'], dem['physician-fee-freeze'], nan_policy='omit')

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

From this test we can see that Republicans supported this bill much more than Democrats. This is easy to see when you look at both the T statistic and p-value. Our T Statistic is far greater than 0 and positive. This would indicate that our mean for 'rep' is much bigger than 'dem'. likewise our miniscule p-value show a big discrepancy between means. 

In [107]:
# Third Test 'water-project'
stats.ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

- The results from this test show very close $\mu$ between Republicans and Democrats on the issue of water projects. Since we can see thT statistic is postive we know that republicans supported this bill more than Democrats. That being said, our p-value is vey close to 1 meaning the differen in mean between the two is very miniscule 

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [108]:
Allvotes = df2[df2.columns[1:]].apply(lambda x: stats.ttest_ind(rep[x], dem[x], nan_policy='omit'))

KeyError: ignored

In [0]:
# try to make a function 
def ttest2sampfunction(contents):
  return stats.ttest_ind(rep[contents], dem[contents], nan_policy='omit')

In [0]:
df2.columns.apply(ttest2sampfunction)

In [0]:
df2[].apply(ttest2sampfunction)

In [0]:
### YOUR WORK HERE
df2_allvotes = stats.ttest_ind(rep['handicapped-infants'], dem['handicapped-infants'], nan_policy='omit')



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





# My sample test will be on the cars dataset.
 I will test perfom a test on city-mpg. We will use a sample size of 100. Our null hypothesis will be that a "cars from the dealership have a $\mu$ of 12 mpg.   

In [110]:
### YOUR WORK HERE
# Making our sample
tsample = df['city-mpg'].sample(n=100, random_state=30)
tsample.describe()

count    100.00000
mean      25.20000
std        6.24904
min       14.00000
25%       20.00000
50%       24.50000
75%       30.00000
max       45.00000
Name: city-mpg, dtype: float64

In [113]:
# cCalculating the T statistic 
t = (25.2-12)/(6.25/np.sqrt(100))
print(t)

21.119999999999997


### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [0]:
### YOUR WORK HERE